### PR TITLE
add empty chart message

### LINKED ISSUE #76: Empty state for chart with no data
When a chart can't be plotted or has no data, the empty state doesn't render anything and the configure panel is quite confusing:

Screenshot 2021-06-20 at 13 18 49

### PR DESCRIPTION
**Pull request by ZeRego**
Closes #76 

- Small change in the hook to assert seriesLayout.yMetrics is not an empty array
- Add message when chart has no data
- Disable chart buttons when chart has no data
- Small typing improvements

Preview:
Screenshot 2021-06-22 at 16 29 42

### CODE REVIEW

**Comment by ZeRego** [comment]
@all-contributors add @Jose Rego for code, design, docs

File: CHANGELOG.md [hunk 1]
```diff
@@ -5,6 +5,7 @@ Recent and upcoming changes to lightdash
 ## Unreleased
 ### Fixed
  - bug where long lists of table were not scrollable in the side bar
+ - handle empty chart
```
Comment (owlas) [inline CHANGELOG.md:8]: The changelog should be really user friendly, so more about what the benefit is to them:

> improved chart UI when there isn't enough data to create a chart

**Comment by ZeRego** [comment]
@all-contributors add @ZeRego for code, design, docs

File: packages/frontend/src/components/ChartConfigPanel.tsx [hunk 2]
```diff
@@ -56,9 +56,10 @@ const Content = ({ chartConfig }: ContentProps) => {
 }
 
 type ChartConfigPanelProps = {
-    chartConfig: ChartConfig,
+    chartConfig: ChartConfig;
+    disabled: boolean;
 }
-export const ChartConfigPanel = ({ chartConfig }: ChartConfigPanelProps) => {
+export const ChartConfigPanel: React.FC<ChartConfigPanelProps> = ({ chartConfig, disabled }) => {
```
Comment (owlas) [inline packages/frontend/src/components/ChartConfigPanel.tsx:62]: Thanks, I'll make this standard everywhere

File: packages/frontend/src/components/Explorer.tsx [hunk 3]
```diff
@@ -56,11 +57,11 @@ export const Explorer = () => {
...
-                            <Button active={activeVizTab === 'scatter' } icon={'scatter-plot'} onClick={() => setActiveVizTab('scatter')}>Scatter</Button>
-                            <ChartConfigPanel chartConfig={chartConfig}/>
+                            <Button active={activeVizTab === 'column' } icon={'timeline-bar-chart'} onClick={() => setActiveVizTab('column')} disabled={isChartEmpty}>Column</Button>
+                            <Button active={activeVizTab === 'bar' } icon={'horizontal-bar-chart'} onClick={() => setActiveVizTab('bar')} disabled={isChartEmpty}>Bar</Button>
+                            <Button active={activeVizTab === 'line' } icon={'timeline-line-chart'} onClick={() => setActiveVizTab('line')} disabled={isChartEmpty}>Line</Button>
+                            <Button active={activeVizTab === 'scatter' } icon={'scatter-plot'} onClick={() => setActiveVizTab('scatter')} disabled={isChartEmpty}>Scatter</Button>
+                            <ChartConfigPanel chartConfig={chartConfig} disabled={isChartEmpty}/>
```
Comment (owlas) [inline packages/frontend/src/components/Explorer.tsx:64]: Isn't this derived from `chartConfig` ? why the extra prop?

File: packages/frontend/src/hooks/useChartConfig.tsx [hunk 4]
```diff
@@ -106,7 +111,7 @@ export const useChartConfig = (queryResults: UseQueryResult<ApiQueryResults, Api
             setXDimension,
             setGroupDimension,
             toggleYMetric,
-            seriesLayout: {xDimension: seriesLayout.xDimension, yMetrics: seriesLayout.yMetrics, groupDimension: seriesLayout.groupDimension},  // verbose to help typescript
+            seriesLayout,
```
Comment (owlas) [inline packages/frontend/src/hooks/useChartConfig.tsx:114]: I can't remember why this was needed, did typing work for you without it?

**Review by owlas — CHANGES_REQUESTED** [review]

File: packages/frontend/src/hooks/useChartConfig.tsx (hunk 4 above)
Comment (ZeRego) [reply packages/frontend/src/hooks/useChartConfig.tsx:114]: The new function `isValidSeriesLayout` has 2 purposes:
- validates data
- confirms the input type ( notice the function return typing: `: seriesLayout is ValidSeriesLayout` )

```const isValidSeriesLayout = (seriesLayout: SeriesLayout): seriesLayout is ValidSeriesLayout => {
    return !!seriesLayout.xDimension && !!seriesLayout.yMetrics && seriesLayout.yMetrics.length > 0;
}```

So if the function returns true Typescript knows that inside the `if` block seriesLayout has type `ValidSeriesLayout` so we can remove that hack.

**Pull request merged**